Description: display the "new document" menu under unity, even if there is no
 template installed
Bug-Ubuntu: https://launchpad.net/bugs/1113648
diff --git a/src/nautilus-actions.h b/src/nautilus-actions.h
index 8c8e575..2b9e642 100644
--- a/src/nautilus-actions.h
+++ b/src/nautilus-actions.h
@@ -119,6 +119,7 @@
 #define NAUTILUS_ACTION_SCRIPTS "Scripts"
 #define NAUTILUS_ACTION_OPEN_SCRIPTS_FOLDER "Open Scripts Folder"
 #define NAUTILUS_ACTION_NEW_DOCUMENTS "New Documents"
+#define NAUTILUS_ACTION_NEW_EMPTY_DOCUMENT "New Empty Document"
 #define NAUTILUS_ACTION_EMPTY_TRASH_CONDITIONAL "Empty Trash Conditional"
 #define NAUTILUS_ACTION_MANUAL_LAYOUT "Manual Layout"
 #define NAUTILUS_ACTION_REVERSED_ORDER "Reversed Order"
diff --git a/src/nautilus-directory-view-ui.xml b/src/nautilus-directory-view-ui.xml
index 64a9fc6..ad98f81 100644
--- a/src/nautilus-directory-view-ui.xml
+++ b/src/nautilus-directory-view-ui.xml
@@ -12,6 +12,7 @@
       <menu action="New Documents">
         <placeholder name="New Documents Placeholder"/>
         <separator name="After New Documents"/>
+        <menuitem name="New Empty Document" action="New Empty Document"/>
       </menu>
     </placeholder>
     <placeholder name="Open Placeholder">
@@ -95,6 +96,8 @@
     <menuitem name="New Folder with Selection" action="New Folder with Selection"/>
     <menu action="New Documents">
       <placeholder name="New Documents Placeholder"/>
+      <separator name="After New Documents"/>
+      <menuitem name="New Empty Document" action="New Empty Document"/>
     </menu>
   </placeholder>
   <placeholder name="Global File Items Placeholder">
@@ -121,6 +124,8 @@
       <menuitem name="New Folder" action="New Folder"/>
       <menu action="New Documents">
         <placeholder name="New Documents Placeholder"/>
+        <separator name="After New Documents"/>
+        <menuitem name="New Empty Document" action="New Empty Document"/>
       </menu>
     </placeholder>
     <separator name="Clipboard separator"/>
diff --git a/src/nautilus-view.c b/src/nautilus-view.c
index 481d039..049260e 100644
--- a/src/nautilus-view.c
+++ b/src/nautilus-view.c
@@ -2133,6 +2133,15 @@ action_new_folder_with_selection_callback (GtkAction *action,
 }
 
 static void
+action_new_empty_file_callback (GtkAction *action,
+				gpointer callback_data)
+{                
+        g_assert (NAUTILUS_IS_VIEW (callback_data));
+
+	nautilus_view_new_file (NAUTILUS_VIEW (callback_data), NULL, NULL);
+}
+
+static void
 action_properties_callback (GtkAction *action,
 			    gpointer callback_data)
 {
@@ -7166,6 +7175,11 @@ static const GtkActionEntry directory_view_entries[] = {
   /* label, accelerator */       N_("New Folder with Selection"), NULL,
   /* tooltip */                  N_("Create a new folder containing the selected items"),
 				 G_CALLBACK (action_new_folder_with_selection_callback) },
+  /* name, stock id */         { NAUTILUS_ACTION_NEW_EMPTY_DOCUMENT, NULL,
+    /* translators: this is used to indicate that a document doesn't contain anything */
+  /* label, accelerator */       N_("_Empty Document"), NULL,
+  /* tooltip */                  N_("Create a new empty document inside this folder"),
+				 G_CALLBACK (action_new_empty_file_callback) },
   /* name, stock id */         { NAUTILUS_ACTION_OPEN, NULL,
   /* label, accelerator */       N_("_Open"), "<control>o",
   /* tooltip */                  N_("Open the selected item in this window"),
@@ -7420,6 +7434,27 @@ static const GtkToggleActionEntry directory_view_toggle_entries[] = {
 };
 
 static void
+connect_proxy (NautilusView *view,
+	       GtkAction *action,
+	       GtkWidget *proxy,
+	       GtkActionGroup *action_group)
+{
+	GdkPixbuf *pixbuf;
+	GtkWidget *image;
+
+	if (strcmp (gtk_action_get_name (action), NAUTILUS_ACTION_NEW_EMPTY_DOCUMENT) == 0 &&
+	    GTK_IS_IMAGE_MENU_ITEM (proxy)) {
+		pixbuf = nautilus_ui_get_menu_icon ("text-x-generic", GTK_WIDGET (view));
+		if (pixbuf != NULL) {
+			image = gtk_image_new_from_pixbuf (pixbuf);
+			gtk_image_menu_item_set_image (GTK_IMAGE_MENU_ITEM (proxy), image);
+
+			g_object_unref (pixbuf);
+		}
+	}
+}
+
+static void
 pre_activate (NautilusView *view,
 	      GtkAction *action,
 	      GtkActionGroup *action_group)
@@ -7482,6 +7517,9 @@ real_merge_menus (NautilusView *view)
 	g_object_unref (action);
 	g_free (tooltip);
 
+	g_signal_connect_object (action_group, "connect-proxy",
+				 G_CALLBACK (connect_proxy), G_OBJECT (view),
+				 G_CONNECT_SWAPPED);
 	g_signal_connect_object (action_group, "pre-activate",
 				 G_CALLBACK (pre_activate), G_OBJECT (view),
 				 G_CONNECT_SWAPPED);
@@ -8813,7 +8851,11 @@ real_update_menus (NautilusView *view)
 	action = gtk_action_group_get_action (view->details->dir_action_group,
 					      NAUTILUS_ACTION_NEW_DOCUMENTS);
 	gtk_action_set_sensitive (action, can_create_files);
-	gtk_action_set_visible (action, !selection_contains_recent && view->details->templates_present);
+	
+	if (g_strcmp0 (g_getenv ("XDG_CURRENT_DESKTOP"), "Unity") == 0)
+		gtk_action_set_visible (action, TRUE);
+	else
+		gtk_action_set_visible (action, !selection_contains_recent && view->details->templates_present);
 
 	action = gtk_action_group_get_action (view->details->dir_action_group,
 					      NAUTILUS_ACTION_COPY_TO);
